drivers/i2c/i2c_ll_stm32.c

```diff
@@ -535,45 +536,88 @@ void i2c_stm32_smbalert_disable(const struct device *dev)
 }
 #endif /* CONFIG_SMBUS_STM32 */
 
+#ifdef I2C_STM32_V2_DMA
+/* This function is executed in the interrupt context */
+static void dma_rx_callback(const struct device *dma_dev, void *user_data, uint32_t channel,
+			    int status)
+{
+	struct i2c_stm32_data *data = (struct i2c_stm32_data *)user_data;
+
+	switch (status) {
+	case DMA_STATUS_COMPLETE:
+		data->current.is_err = 0;
+		dma_stop(dma_dev, channel);
+		break;
+	case DMA_STATUS_BLOCK:
+		__ASSERT(false, "Unsupported feature");
+		break;
+	default:
+		data->current.is_err = 1;
+		break;
+	}
+
+	k_sem_give(&data->device_sync_sem);
+}
+
+/* This function is executed in the interrupt context */
+static void dma_tx_callback(const struct device *dma_dev, void *user_data, uint32_t channel,
+			    int status)
+{
+	struct i2c_stm32_data *data = (struct i2c_stm32_data *)user_data;
+
+	switch (status) {
+	case DMA_STATUS_COMPLETE:
+		data->current.is_err = 0;
+		dma_stop(dma_dev, channel);
+		break;
+	case DMA_STATUS_BLOCK:
+		__ASSERT(false, "Unsupported feature");
+		break;
+	default:
+		data->current.is_err = 1;
+		break;
+	}
+
+	LOG_DBG("Giving semaphore");
+	k_sem_give(&data->device_sync_sem);
+}
+
+#endif /* I2C_STM32_V2_DMA */
+
```

Comment:
I strongly suggest to remove both DMA interrupts
They  only complicates the implementation, or maybe i have missed something?
Better to use the I2C flags to determine when data->device_sync_sem shall be given

Reply:
Added to catch potential DMA transfer errors and return the taken semaphore, also followed either SPI or UART (I forget which) as an example and it was used there. Feels a bit belt and braces now, but may be useful for async support?

Reply:
Instead of DMA_STATUS_COMPLETE, it is better to use I2C TC flag as it was before
Beacuse when DMA transfer is complete in Write mode, the last byte is not yet transmitted by I2C peripheral
so it is to early to give device_sync_sem in dma callback

DMA errors could be catched by checking DMA status after k_sem_take(&data->device_sync_sem,...) is done

Reply:
Happy to take guidance on this but my feeling is we want to have a basic level of isr handling for the dma to handle dma events cleanly, otherwise the i2c isr may balloon to cope

Reply:
Should be no differences in i2c isr with/without the DMA callbacks

For other peripherals which don't have a built-in transfer counter, DMA callbacks must be used to swap buffers, notify upper layers etc. 

But in this case we already have a byte counter in the i2c peripheral which trigger interrupts when all data is transferred and sets ISR flags TC and TCR.

My point is that it should just be one of the counters when reaching 0 that should trigger device_sync_sem and i argue that the I2C flags are the natural choice in this case

Reply:
Good point well argued but what about a DMA fault/error condition? I would argue that having these callbacks we can more easily implement and distinguish how we handle those particular events.

Looking at the stm32 LL driver, we might be able to use these callbacks to get around the 255 byte transfer limit to load the next chunk without going through the whole init process. I recognise the work hasn't been done but it lays the ground work for very little overhead

Reply:
For the errors i can see two cases:
1. Configuration errors for example data alignement, src/dest bus witdth:s or src/dest pointers are setup wrong. This should be no problem here as we only work with bytes and validating pointers can (and should) be done before starting a transaction
2. Bus starvation, for example when receiving with high speed and destination is in external SDRAM which might already be very busy with other transactions. This has happened to me at rare occasions when i was reading from ADC at high speeds on a low end STM32
I consider this a design fault, In the same way as we must ensure the CPU has enough processing power to meet SW deadlines, the bus bandwidth allocation must also be tuned to fit the application

IMO if we can consider DMA faults as design faults that should be fixed (in contrary to I2C faults which can be caused by noise on I2C lines for example) then i would say its better to keep the complexity down, skipping the DMA callbacks and just checking the DMA flags when transaction finishes/times out
But maybe there are more error causes than the two i listed, that really needs to be handled in a dma callback which, if so i rest my case ... :)

Regarding the 255 byte limit and Reload of I2C counter i agree that the current implementation is somewhat inefficient, as it needs to wakeup thread to make reload of the I2C counter
In my PR https://github.com/zephyrproject-rtos/zephyr/pull/82940
i have improved that part so that the reload is done within the I2C ISR instead

Reply:
OK removed the callbacks but I wonder if we will need them sooner or later but we can burn that bridge when we get there